libs/framework/src/bundle_archive.c

```diff
@@ -28,644 +28,415 @@
-			tm_time.tm_year = year - 1900;
-			tm_time.tm_mon = month - 1;
-			tm_time.tm_mday = day;
-			tm_time.tm_hour = hours;
-			tm_time.tm_min = minutes;
-			tm_time.tm_sec = seconds;
-
-			*time = mktime(&tm_time);
-		}
-	}
-
-	framework_logIfError(archive->fw->logger, status, NULL, "Could not read last modified");
-
-	return status;
+celix_status_t celix_bundleArchive_getLastModified(bundle_archive_pt archive, struct timespec* lastModified) {
+    celix_status_t status;
+    celixThreadMutex_lock(&archive->lock);
+    status = celix_bundleArchive_getLastModifiedInternal(archive, lastModified);
+    celixThreadMutex_unlock(&archive->lock);
+    return status;
 }
 
-static celix_status_t bundleArchive_writeLastModified(bundle_archive_pt archive) {
-	celix_status_t status = CELIX_SUCCESS;
-	FILE *lastModifiedFile;
-	char lastModified[512];
-
-	snprintf(lastModified, sizeof(lastModified), "%s/bundle.lastmodified", archive->archiveRoot);
-	lastModifiedFile = fopen(lastModified, "w");
-	if (lastModifiedFile == NULL) {
-		status = CELIX_FILE_IO_EXCEPTION;
-	} else {
-		char timeStr[20];
-		strftime(timeStr, 20, "%Y %m %d %H:%M:%S", localtime(&archive->lastModified));
-		fprintf(lastModifiedFile, "%s", timeStr);
-		fclose(lastModifiedFile);
-	}
-
-	framework_logIfError(celix_frameworkLogger_globalLogger(), status, NULL, "Could not write last modified");
-
-	return status;
+celix_status_t bundleArchive_setLastModified(bundle_archive_pt archive __attribute__((unused)), time_t lastModifiedTime  __attribute__((unused))) {
+    return celix_utils_touch(archive->archiveRoot);
 }
 
-celix_status_t bundleArchive_revise(bundle_archive_pt archive, const char * location, const char *inputFile) {
-	celix_status_t status = CELIX_SUCCESS;
-	long revNr = 0l;
-	if (!linkedList_isEmpty(archive->revisions)) {
-		long revisionNr;
-		status = bundleRevision_getNumber(linkedList_getLast(archive->revisions), &revisionNr);
-		revNr = revisionNr + 1;
-	}
-	if (status == CELIX_SUCCESS) {
-		status = bundleArchive_reviseInternal(archive, false, revNr, location, inputFile);
-	}
-
-	framework_logIfError(celix_frameworkLogger_globalLogger(), status, NULL, "Could not revise bundle archive");
-
-	return status;
+celix_status_t bundleArchive_revise(bundle_archive_pt archive, const char * location, const char *updatedBundleUrl) {
+    celixThreadMutex_lock(&archive->lock);
+
+    const char* updateUrl = archive->location;
+    if (updatedBundleUrl != NULL && strcmp(archive->location, updatedBundleUrl) != 0) {
+        fw_log(archive->fw->logger, CELIX_LOG_LEVEL_INFO, "Updating bundle archive bundle url location to %s", updatedBundleUrl);
+        updateUrl = updatedBundleUrl;
+    }
+
+    struct timespec lastModified;
+    celix_bundleArchive_getLastModifiedInternal(archive, &lastModified);
+    celix_status_t status = celix_bundleArchive_extractBundle(archive, archive->resourceCacheRoot, updateUrl, &lastModified);
```

Comment:
If `celix_bundleArchive_extractBundle` failed due to broken zip file, then this bundle is totally broken. For the original multi-revision design (though with broken implementation), we don't have such issue.